includes/class-wc-payments-order-success-page.php

```diff
@@ -14,11 +14,33 @@ class WC_Payments_Order_Success_Page {
 	 * Constructor.
 	 */
 	public function __construct() {
+		add_filter( 'woocommerce_thankyou_order_id', [ $this, 'update_order_meta' ] );
 		add_filter( 'woocommerce_thankyou_order_received_text', [ $this, 'show_woopay_thankyou_notice' ], 10, 2 );
 		add_action( 'woocommerce_thankyou_woocommerce_payments', [ $this, 'show_woopay_payment_method_name' ] );
 		add_action( 'wp_enqueue_scripts', [ $this, 'enqueue_scripts' ] );
 	}
 
+	public function update_order_meta( $order_id ) {
+		if ( $order_id <= 0 ) {
+			return $order_id;
+		}
+
+		$order = wc_get_order( $order_id );
+
+		if ( ! $order ) {
+			return $order_id;
+		}
+
+		if ( ! empty( $_SERVER['HTTP_REFERER'] ) && wp_unslash( $_SERVER['HTTP_REFERER'] ) ?? '' !== 'http://woopay.test' ) {
```

Comment:
I'm very unhappy with how we're detecting a request coming from WooPay, but can't really think of a different way to do this. We at least need some way to allow both the production URL as well as our local development environment URLs.

Then there's the question of how sites behind a reverse proxy or load balancer may potentially alter the value of this variable, so it doesn't feel very reliable. I'll be looking at ways to improve this, but would love to hear some feedback or ideas on this or other approaches.

Reply:
We could add a query parameter to the URL for the Order Received page redirect but then anyone can do that and trigger this code path. Not sure if that's the right approach here either.

Reply:
Resolving this since the approach has changed significantly since I initially worked on this.